bin/propolis-server/src/lib/spec/api_request.rs

```diff
@@ -152,3 +152,59 @@ pub(super) fn parse_nic_from_request(
         backend_spec,
     })
 }
+
+#[cfg(test)]
+mod test {
+    use propolis_api_types::VolumeConstructionRequest;
+    use uuid::Uuid;
+
+    use super::*;
+
+    fn check_parsed_storage_device_backend_pointer(
+        parsed: &ParsedStorageDevice,
+    ) {
+        let device_to_backend = match &parsed.device_spec {
+            StorageDeviceV0::VirtioDisk(d) => d.backend_name.clone(),
+            StorageDeviceV0::NvmeDisk(d) => d.backend_name.clone(),
+        };
+
+        assert_eq!(device_to_backend, parsed.backend_name);
```

Comment:
it seems like instead of having `ParsedStorageDevice.backend_name` and requiring it always matches what happens to also be in the device spec, that we'd be better off with a `StorageDeviceV0::backend_name()` helper that we use in places like `add_storage_device()`.  something like `.insert(device_spec.backend_name(), backend_spec)`?

do we expect to have storage devices that don't know their backend name? (that seems unlikely, but it's an obvious reason my thought wouldn't stick!)

(i also understand this change and why it fixes things, so +1 for landing to make things fixed and adjusting after if you're so inclined)

Reply:
Ultimately my plan is to do away with `ParsedStorageDevice` entirely and replace it with a `Disk` type that contains both the device and backend halves of each of a VM's storage components. In this bug I got caught looking the wrong way halfway through the series of refactoring PRs that's supposed to make that happen ;)

I have the result of this wave of refactoring in [a branch in my fork](https://github.com/oxidecomputer/propolis/compare/master...gjcolombo:propolis:instance-spec-rework#diff-f8258ff7a3e14ba17fbb2c5c6655bea413975c40642622e5eeb3ba39cfa03729); here's what [the routine this PR fixes](https://github.com/gjcolombo/propolis/blob/2f65e8cd8f0d82ac09573d2a5edcf217798f9a49/bin/propolis-server/src/lib/spec/api_request.rs#L71-L102) will look like when that's done. This will (hopefully) foreclose on the sort of bug we hit here, since machine init won't be looking up named backends in a map anymore. (I think the code can be improved even more beyond what I have in the fork at this point, but am trying to stick to one renovation project at a time--insert the beaver-big-plans meme here...) 

There's even more context in #735 if you're looking for some light bedtime reading.